### ISSUE TITLE
Redis: Issue with GET command 

### ISSUE THREAD

**Issue by 0fca** [boazsegev/facil.io#157]
# Issue
Hi there. 
Do have in mind I can be solely mistaken about all the case, but I cannot find the workaround or the solution, thus I thought I will create this issue. I am not very experienced with C or low-level Redis access.
I have met an issue with GET command.
I mean, the actual problem seems to be persistent - whenever I log out Redis response from `reply` object using `fprintf` it always returns a string similar to that one:
```
WARNING: (redis 37900) received a reply when no command was sent.
-ERR unknown command `$11`, with args beginning with
````
And my callback function looks like this:
```
void redis_callback(fio_pubsub_engine_s* engine, FIOBJ reply, void *udata)
{
    if(reply == FIOBJ_INVALID)
    {
        fprintf(stderr, "It appears that there was an error on Redis op");
    }

    fprintf(stderr, fiobj_obj2cstr(reply).data);
}
```
However, in case of simple `SET` with both key and value strings it just sets the value in the cache and all done.
But whenever I need to `GET` something I just receive the line above and no value from cache. Even though I am sure that the key is correct and it is in the cache. The value is there too. I managed to add a `fprintf` inside `redis_engine.c` file and it printed me twice or more time something about that:
```
$11
GET key
```
Assuming that my key is named key in Redis, of course.
So, to sum up, the issue in my case is about `GET` cause I cannot read values from cache, but it seems that it appears any time.

I use `redis_engine_send` function to call `SET` and `GET` commands to Redis server.

# Setup
facil: 0.7.6
Redis: 5.0.14.1, Windows 11/7.0.2, Debian 11, kernel: 5.15 (barebone, its not the same as build platform)
Build platform: Debian 11, WSL2, kernel: 5.15
gcc: 12.2.0
C compatibility: C-11

# Possible cause

I suspect that the problem is somewhere inside the parser of redis commands.
I think so 'cause the problem occurs every time I send the command to the server.

**Comment by 0fca** [comment]
Okay, I have fiddled a little with Redis via telnet connection. 
Mentioned `$xx` is just a type of data which is returned with its character length. 
Thus, I dont know if this is something on my part or is it actually something with resp_parser?

**Comment by boazsegev** [comment]
Hi @0fca ,

I'll try to be short as I have little time. If it's still unclear, please write me and I'll try again later.

Usually this error occurs when more responses are received than requests are sent. **This could happen if your `FIOBJ` command contains too much data or more than one command**.

The `redis` engine manages each command, validates the sending (in case of network interruption it will try the command again) and sends back the response to the callback.

If there are more commands than callbacks, it will error out, not knowing where to send the response. That's the logged error message you're seeing.

I will need more code to see where and what went wrong. Hopefully this helps enough so you can solve this yourself.

Good Luck!

**Comment by 0fca** [comment]
Hi there.
Thank you very much for your time. I do appreciate it very much.
I will fiddle with it a little bit. I will check out whether is it a size of data block I try to save or is it something with format/binary structure of command string I encode into `FIOBJ` command object.
I will give you feedback despite its result - either will it be a success or failure.
Thanks again.
Thanmk you for your astonishing work with facil.

**Comment by boazsegev** [comment]
FYI: I assume you are aware that the FIOBJ command should be basically an Array of Strings (or a single String). No need for Redis formatting of the command object.

The library handles the formatting part: https://github.com/boazsegev/facil.io/blob/f261935c6c8fc56d582f0a95e3992bcb324a1f26/lib/facil/redis/redis_engine.c#L175

**Comment by 0fca** [comment]
Yes, I do know, I took a look at it.
How I do try to send the `GET` command looks like that:
```c
    int c_count = 6 + strlen(sess_id_raw);
    char* get_command_str = malloc(c_count * sizeof(char));
    snprintf(get_command_str, c_count, "%s %s", "GET", sess_id_raw);
    fprintf(stderr, get_command_str);
    FIOBJ get_command = fiobj_str_new(get_command_str, strlen(get_command_str));
    redis_engine_send(FIO_PUBSUB_DEFAULT, get_command, on_get_command, NULL);
```
The error I get on the `on_get_command` callback is actually a length of a string with `GET` itself.

**Comment by boazsegev** [comment]
Thank you for the example, I think this helps narrow down the issue.

I assume what you wanted was more on the lines of (sorry for the pseudo code rather than actual code, as I'm typing into GitHub):

```c
FIOBJ cmd = fiobj_ary_new();
fio_ary_push(cmd, fiobj_str_new("GET", 3));
fio_ary_push(cmd, fiobj_str_new(sess_id_raw, strlen(sess_id_raw)));
redis_engine_send(FIO_PUBSUB_DEFAULT, cmd, on_get_command, NULL);
fiobj_free(cmd);
```

After all, in your example you have one long command string "GET ID_XXXX" rather than two command strings "GET" "ID_XXXX".

**Comment by 0fca** [comment]
Thank you for your time.
It seems to work now after I changed it all to use FIOBJ Array.
However, I wonder why would single String cause all that fuss.
What's more it is not stated in the docs that the user shall only use FIOBJ Array object, is it?
If not, I could make a fix for it in the docs to add a note about the command object.
